SAMPLE_FORMAT.md

```diff
@@ -25,19 +57,149 @@ specifically mentioned in the format, samples should make best-effort attempts i
+failures) as part of the test suite. Please note that tests should run successfully in parallel, 
+and UUIDs should be used to prevent conflicts.
+
+Snippets should have integration tests that should verify the snippet compiles and runs
+successfully. Tests should only verify that the sample itself is interacting with the service
+correctly - it is an explicit non goal for tests to verify that API is performing correctly.
+Because of this, mocks for external services are strongly discouraged.
+
+* Test Library: [JUnit4](https://junit.org/junit4/)
+* Test Runner: [Maven Failsafe plugin](https://maven.apache.org/surefire/maven-failsafe-plugin/)
+and [Maven Surefire plugin](https://maven.apache.org/surefire/maven-surefire-plugin/).
+
+As an example, the following test code shows how we test a region tag (region tags are tools Google
+uses to identify sections of the snippets to be highlighted in documentation) called `region_tag`:
+```java
+package com.google.example;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Store region_tag in test method name, camel-cased
+ */ 
+@RunWith(JUnit4.class)
+public class SomeClassTest {
+  @Test public void regionTag_shouldPass() throws Exception { }
+
+  @Test public void regionTag_shouldFail() throws Exception {
+    Assert.fail("should fail");
+  }
+}
+```
+You will note:
+ * underscores (`_`) are used in test method names to separate blocks of `camelCase` text
+ * these blocks denote _region tags_, which serve as unique IDs for snippets in a given repository
+ * `camelCase` blocks beginning with `should` or `does` are **not** region tags - instead, they
+ simply describe the test being run
+
+
+that the "_" in `region_tag` is removed, and "_" is used to separate regionTags from
+test descriptions.
+
+It is also possible to use annotations to provide info for `region_tag` if you need to do this,
+please contact one of the repo admins.
+
+### Testing Setup
+Most samples require a GCP project and billing account. Keep the following in
+mind when setting up tests.
+
+* **Environment variables**  
+  Minimize additional environment variables that need to be set to run the tests.
+  If you do require additional environment variables, they should be added to
+  `run_tests.sh`.
+
+  Existing environment variables include:
+  * `GOOGLE_APPLICATION_CREDENTIALS`
+  * `GOOGLE_CLOUD_PROJECT`
+  * `PROJECT_ID`
+
+* **API library**  
+  If an API needs to be enabled in the testing project, add this information to the
+  directory's README and to the comments in the PR. If there is no README.md file, add one
+  in your PR.
+
+* **IAM**
+  Some API's require that the service account have some additional capibilities, these should also
+  be mentioned in both the PR and the README.
+
+* **Cloud resources**  
+  Most Java samples create the Cloud resources that they need to run. If this
```

Comment:
As per the "keeping tests from getting stale" lightning talk we recently had, might want to add a note about how tests that create cloud resources should also delete those resources when they are done testing, particularly in a way that will ensure the deletion of the resource even if the test fails, such as with with a `finally` block or in an `@After` or `@AfterClass` function.

Reply:
done